internal/controller/controllers/installenv_controller.go

```diff
@@ -0,0 +1,218 @@
+	corev1 "k8s.io/api/core/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	"github.com/openshift/assisted-service/internal/bminventory"
+	"github.com/openshift/assisted-service/internal/common"
+	adiiov1alpha1 "github.com/openshift/assisted-service/internal/controller/api/v1alpha1"
+	"github.com/openshift/assisted-service/models"
+	"github.com/openshift/assisted-service/restapi/operations/installer"
+)
+
+const (
+	ImageCreated       = "ImageCreated"
+	ImageBeingCreated  = "ImageBeingCreated"
+	ImageCreationError = "ImageCreationError"
+)
+
+type ImageState struct {
+	State     string
+	StateInfo string
+}
+
+// InstallEnvReconciler reconciles a InstallEnv object
+type InstallEnvReconciler struct {
+	client.Client
+	Log                      logrus.FieldLogger
+	Installer                bminventory.InstallerInternals
+	PullSecretUpdatesChannel chan event.GenericEvent
+}
+
+// +kubebuilder:rbac:groups=adi.io.my.domain,resources=installenvs,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=adi.io.my.domain,resources=installenvs/status,verbs=get;update;patch
+
+func (r *InstallEnvReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	ctx := context.Background()
+
+	installEnv := &adiiov1alpha1.InstallEnv{}
+	if err := r.Get(ctx, req.NamespacedName, installEnv); err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	return r.generate(ctx, installEnv)
+}
+
+func (r *InstallEnvReconciler) generate(ctx context.Context, installEnv *adiiov1alpha1.InstallEnv) (ctrl.Result, error) {
+	var inventoryErr error
+	var imageInfo *models.ImageInfo
+	var updatedCluster *common.Cluster
+	imageState := ImageState{
+		State:     ImageCreationError,
+		StateInfo: adiiov1alpha1.ImageStateFailedToCreate,
+	}
+
+	clusterDeployment, clusterDeploymentRefErr := r.getClusterDeploymentByRef(ctx, installEnv.Spec.ClusterRef)
+	if clusterDeploymentRefErr != nil {
+		return r.updateStatusAndReturnResult(ctx, installEnv, nil, imageState, clusterDeploymentRefErr)
+	}
+
+	c, err := r.Installer.GetClusterByKubeKey(types.NamespacedName{
+		Name:      clusterDeployment.Name,
+		Namespace: clusterDeployment.Namespace,
+	})
+
+	if gorm.IsRecordNotFoundError(err) {
+		inventoryErr = common.NewApiError(http.StatusNotFound, err)
+
+	} else {
+		isoParams := installer.GenerateClusterISOParams{
+			ClusterID:         *c.ID,
+			ImageCreateParams: &models.ImageCreateParams{},
+		}
+		if len(installEnv.Spec.SSHAuthorizedKeys) > 0 {
+			isoParams.ImageCreateParams.SSHPublicKey = installEnv.Spec.SSHAuthorizedKeys[0]
+		}
+		updatedCluster, inventoryErr = r.Installer.GenerateClusterISOInternal(ctx, isoParams)
```

Comment:
Does this always generate a new ISO? Or it only does so if the parameters have changed? It would be useful to add a comment here either way. Controllers often have to handle that logic of only calling some function when its input has changed, so if that problem is solved here in the function itself and can be omitted from the controller, that's helpful to state.

Reply:
Reconcile functions should be able to run (usually) quickly and many times in rapid succession. Normally you would not want a reconcile function to re-generate an ISO every single time it runs. This function call sort of looks like it's doing that, which would cause concern for someone familiar with controllers but not familiar with that function. Since you and I have internal knowledge that calling this function named "create ISO" won't actually create an ISO most of the time, it would be useful to future readers to add a brief comment about that IMO.